vaadin-tabs-flow-parent/vaadin-tabs-flow/src/main/java/com/vaadin/flow/component/tabs/TabSheet.java

```diff
@@ -0,0 +1,291 @@
+import com.vaadin.flow.component.ComponentEvent;
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.component.HasSize;
+import com.vaadin.flow.component.HasStyle;
+import com.vaadin.flow.component.Tag;
+import com.vaadin.flow.component.Text;
+import com.vaadin.flow.component.dependency.JsModule;
+import com.vaadin.flow.component.dependency.NpmPackage;
+import com.vaadin.flow.component.shared.HasPrefixAndSuffix;
+import com.vaadin.flow.component.shared.HasThemeVariant;
+import com.vaadin.flow.dom.Element;
+import com.vaadin.flow.shared.Registration;
+
+/**
+ * TabSheet consists of a set of tabs and the content area. The content area
+ * displays a component associated with the selected tab.
+ *
+ * @author Vaadin Ltd.
+ */
+@Tag("vaadin-tabsheet")
+@NpmPackage(value = "@vaadin/tabsheet", version = "23.3.0-alpha1")
+@JsModule("@vaadin/tabsheet/src/vaadin-tabsheet.js")
+public class TabSheet extends Component implements HasStyle, HasSize,
+        HasPrefixAndSuffix, HasThemeVariant<TabSheetVariant> {
+
+    private Tabs tabs = new Tabs();
+
+    private Map<Tab, Element> tabToContent = new HashMap<>();
+
+    /**
+     * The default constructor.
+     */
+    public TabSheet() {
+        super();
+
+        tabs.getElement().setAttribute("slot", "tabs");
+        getElement().appendChild(tabs.getElement());
+
+        addSelectedChangeListener(e -> updateContent());
+    }
+
+    /**
+     * Adds a tab with the given content.
+     *
+     * @param tab
+     *            the tab
+     * @param content
+     *            the content related to the tab
+     * @return the added tab
+     */
+    public Tab add(Tab tab, Component content) {
+        Objects.requireNonNull(tab, "The tab to be added cannot be null");
+        Objects.requireNonNull(content,
+                "The content to be added cannot be null");
+        tabs.add(tab);
+
+        // On the client, content is associated with a tab by id
+        var id = "tabsheet-tab-" + UUID.randomUUID().toString();
+        tab.setId(id);
+        content.getElement().setAttribute("tab", id);
+
+        tabToContent.put(tab, content.getElement());
+
+        updateContent();
+
+        return tab;
+    }
+
+    /**
+     * Adds a tab created from the given tab content and content.
+     *
+     * @param tabContent
+     *            the content of the tab
+     * @param content
+     *            the content related to the tab
+     * @return the created tab
+     */
+    public Tab add(Component tabContent, Component content) {
+        return add(new Tab(tabContent), content);
+    }
+
+    /**
+     * Adds a tab created from the given text and content.
+     *
+     * @param tabText
+     *            the text of the tab
+     * @param content
+     *            the content related to the tab
+     * @return the created tab
+     */
+    public Tab add(String tabText, Component content) {
+        return add(new Text(tabText), content);
```

Comment:
nit: would it make sense to re-order the methods like: 

* add(String tabText, Component content)
* add(Component tabContent, Component content)
* add(Tab tab, Component content)

so that it's easier to see what method is calling the other (string -> new Text -> component -> new Tab -> tab)

Reply:
Done